File: modules/image_generator.py
# modules/image_generator.py
# Playwright automation for Google Flow image generation.

import logging
import time
from playwright.sync_api import Page, Download
from config import (
    FLOW_URL,
    FLOW_PROMPT_INPUT_SELECTOR,
    FLOW_GENERATE_BTN_SELECTOR,
    FLOW_DONE_INDICATOR,
    FLOW_DOWNLOAD_BTN_SELECTOR,
    ELEMENT_TIMEOUT,
    GENERATION_TIMEOUT,
    DOWNLOAD_TIMEOUT,
)
from .downloader import Downloader

logger = logging.getLogger("bot")


class ImageGenerator:
    """
    Controls the Google Flow tab to:
      1. Navigate to the persistent project URL
      2. Type the image prompt
      3. Click Generate
      4. Wait for completion
      5. Download the latest generated image
    """

    def __init__(self, page: Page):
        self.page = page

    def generate(self, prompt_id: int, image_prompt: str) -> str:
        """
        Run the full image generation flow.
        Returns the saved image file path on success.
        Raises an exception on failure.
        """
        logger.info(f"[ID:{prompt_id}] Starting image generation on Google Flow")

        # Step 1: Navigate to the project (persistent session reused)
        self.page.goto(FLOW_URL, wait_until="domcontentloaded")
        logger.debug(f"[ID:{prompt_id}] Navigated to Flow project URL")

        # Step 2: Wait for prompt input to appear
        self.page.wait_for_selector(
            FLOW_PROMPT_INPUT_SELECTOR,
            timeout=ELEMENT_TIMEOUT
        )

        # Step 3: Clear any existing text and type the prompt
        prompt_input = self.page.query_selector(FLOW_PROMPT_INPUT_SELECTOR)
        prompt_input.click()
        prompt_input.fill("")           # Clear existing
        prompt_input.fill(image_prompt) # Type prompt
        logger.debug(f"[ID:{prompt_id}] Prompt typed: {image_prompt[:60]}...")

        # Step 4: Click Generate
        generate_btn = self.page.wait_for_selector(
            FLOW_GENERATE_BTN_SELECTOR,
            timeout=ELEMENT_TIMEOUT
        )
        generate_btn.click()
        logger.info(f"[ID:{prompt_id}] Generate button clicked, waiting for result...")

        # Step 5: Wait for the generated image to appear
        self.page.wait_for_selector(
            FLOW_DONE_INDICATOR,
            timeout=GENERATION_TIMEOUT
        )
        logger.info(f"[ID:{prompt_id}] Image generation complete")

        # Step 6: Download the most recently generated image
        image_path = self._download_latest_image(prompt_id)
        return image_path

    def _download_latest_image(self, prompt_id: int) -> str:
        """
        Click the download button on the latest generated image and save it.
        Google Flow shows the most recent generation last in the grid.
        """
        # Get all download buttons and click the last one (most recent)
        download_btns = self.page.query_selector_all(FLOW_DOWNLOAD_BTN_SELECTOR)
        if not download_btns:
            raise RuntimeError(f"[ID:{prompt_id}] No download button found after generation")

        latest_btn = download_btns[-1]  # Last = most recently generated

        # Capture the download event
        with self.page.expect_download(timeout=DOWNLOAD_TIMEOUT) as download_info:
            latest_btn.click()

        download: Download = download_info.value
        image_path = Downloader.save_image(download, prompt_id)
        return image_path

File: src/imageGenerator.ts
// src/imageGenerator.ts
// Playwright automation for Google Flow image generation.

import { Page } from 'playwright';
import {
  FLOW_URL,
  FLOW_PROMPT_INPUT_SELECTOR,
  FLOW_GENERATE_BTN_SELECTOR,
  FLOW_DONE_INDICATOR,
  FLOW_DOWNLOAD_BTN_SELECTOR,
  FLOW_MODAL_DOWNLOAD_BTN,
  FLOW_QUALITY_1K_BTN,
  FLOW_BACK_BTN_SELECTOR,
  ELEMENT_TIMEOUT,
  GENERATION_TIMEOUT,
  DOWNLOAD_TIMEOUT,
} from './config';
import { Downloader } from './downloader';
import logger from './logger';

export class ImageGenerator {
  private page: Page;

  constructor(page: Page) {
    this.page = page;
  }

  async generate(id: number, imagePrompt: string): Promise<string> {
    logger.info(`[ID:${id}] Starting image generation on Google Flow`);

    // Bring the tab to the foreground
    await this.page.bringToFront();

    // Step 1: Navigate to persistent project URL (only if not already there)
    if (this.page.url() !== FLOW_URL) {
      logger.debug(`[ID:${id}] Navigating to: ${FLOW_URL}`);
      await this.page.goto(FLOW_URL, { waitUntil: 'domcontentloaded' });
      logger.info(`[ID:${id}] Page loaded: ${this.page.url()}`);
    } else {
      logger.info(`[ID:${id}] Already on Flow page, skipping navigation.`);
    }


    // Step 2: Wait for prompt input
    logger.debug(`[ID:${id}] Waiting for prompt input: ${FLOW_PROMPT_INPUT_SELECTOR}`);
    await this.page.waitForSelector(FLOW_PROMPT_INPUT_SELECTOR, {
      timeout: ELEMENT_TIMEOUT,
    });
    logger.info(`[ID:${id}] Found prompt input!`);

    // Step 3: Fill the prompt
    const promptInput = await this.page.$(FLOW_PROMPT_INPUT_SELECTOR);
    if (!promptInput) throw new Error(`[ID:${id}] Prompt input selector failed to return element`);

    logger.info(`[ID:${id}] Clicking and typing into Slate editor...`);
    await promptInput.click();

    // Clear existing text first (Control+A + Backspace)
    await this.page.keyboard.press('Control+A');
    await this.page.keyboard.press('Backspace');

    // Type the new prompt with random human-like speed (40-120ms per char)
    const typingDelay = Math.random() * 80 + 40;
    await this.page.keyboard.type(imagePrompt, { delay: typingDelay });
    logger.info(`[ID:${id}] Success: Prompt typed with ${Math.round(typingDelay)}ms speed.`);



    // Step 4: Click Generate (but count images first!)
    const initialImageCount = (await this.page.$$(FLOW_DONE_INDICATOR)).length;
    logger.info(`[ID:${id}] Current image count: ${initialImageCount}. Waiting for a new one...`);

    logger.debug(`[ID:${id}] Waiting for Generate button: ${FLOW_GENERATE_BTN_SELECTOR}`);
    const generateBtn = await this.page.waitForSelector(
      FLOW_GENERATE_BTN_SELECTOR,
      { timeout: ELEMENT_TIMEOUT }
    );
    logger.info(`[ID:${id}] Found Generate button! Clicking now...`);
    await generateBtn!.click();

    // Step 5: Wait for a NEW generated image to appear
    logger.info(`[ID:${id}] Generate clicked. Waiting up to ${GENERATION_TIMEOUT / 1000}s for a new image...`);

    // Custom wait: Wait until the count of images increases
    await this.page.waitForFunction(
      "({ selector, initialCount }) => document.querySelectorAll(selector).length > initialCount",
      { selector: FLOW_DONE_INDICATOR, initialCount: initialImageCount },
      { timeout: GENERATION_TIMEOUT }
    );



    const finalImageCount = (await this.page.$$(FLOW_DONE_INDICATOR)).length;
    logger.info(`[ID:${id}] New image detected! (Total: ${finalImageCount}).`);

    logger.info(`[ID:${id}] Waiting 70 seconds for image to stabilize before download...`);
    await this.page.waitForTimeout(70_000); 
    logger.info(`[ID:${id}] Wait complete. Starting download.`);



    // Step 6: Download latest image
    return await this.downloadLatestImage(id);
  }



  private async downloadLatestImage(id: number): Promise<string> {
    logger.info(`[ID:${id}] Step 1: Clicking the newest image tile (index 0)...`);
    const tiles = await this.page.$$(FLOW_DONE_INDICATOR);
    if (!tiles.length) throw new Error(`[ID:${id}] No image tiles found to click.`);

    // Click the FIRST tile (newest generation appears at the top/start)
    await tiles[0].click();


    logger.info(`[ID:${id}] Step 2 & 3: Waiting for modal and clicking Download button...`);
    const modalDownloadBtn = await this.page.waitForSelector(FLOW_MODAL_DOWNLOAD_BTN, {
      timeout: ELEMENT_TIMEOUT,
    });
    await modalDownloadBtn!.click();

    logger.info(`[ID:${id}] Step 4 & 5: Selecting 1K quality...`);
    const qualityBtn = await this.page.waitForSelector(FLOW_QUALITY_1K_BTN, {
      timeout: ELEMENT_TIMEOUT,
    });

    logger.info(`[ID:${id}] Clicking 1K and starting actual download...`);
    const [download] = await Promise.all([
      this.page.waitForEvent('download', { timeout: DOWNLOAD_TIMEOUT }),
      qualityBtn!.click(),
    ]);

    const path = await Downloader.saveImage(download, id);
    logger.info(`[ID:${id}] SUCCESS: Image saved to: ${path}`);

    logger.info(`[ID:${id}] Clicking Back to return to project view...`);
    const backBtn = await this.page.waitForSelector(FLOW_BACK_BTN_SELECTOR, {
      timeout: ELEMENT_TIMEOUT,
    });
    await backBtn!.click();

    return path;

  }


}
